File: modules/sql/tables/pjsk.py
from sqlalchemy import (
    Column,
    Integer,
    String,
    Boolean,
    DateTime,
    BigInteger,
    ForeignKey,
    UniqueConstraint
)
from sqlalchemy.orm import relationship

from .base import PjskBase


class UserBinding(PjskBase):
    __tablename__ = "user_bindings"
    __table_args__ = (
        UniqueConstraint("platform", "im_id", "server", name="uq_user_binding"),
    )

    id = Column(Integer, primary_key=True, autoincrement=True)
    platform = Column(String(20), nullable=False)
    im_id = Column(String(30), nullable=False, index=True)
    user_id = Column(String(30), nullable=False)
    server = Column(String(2), nullable=False)
    visible = Column(Boolean, default=True)

    default_refs = relationship(
        "UserDefaultBinding",
        back_populates="binding",
        cascade="all, delete",
        lazy="joined",
    )


class UserDefaultBinding(PjskBase):
    __tablename__ = "user_default_bindings"
    __table_args__ = (
        UniqueConstraint("im_id", "platform", "server", name="uq_user_default_binding"),
    )

    id = Column(Integer, primary_key=True, autoincrement=True)
    im_id = Column(String(30), nullable=False)
    platform = Column(String(20), nullable=False)
    server = Column(String(7), nullable=False)
    binding_id = Column(Integer, ForeignKey("user_bindings.id", ondelete="CASCADE"), nullable=False)

    binding = relationship("UserBinding", back_populates="default_refs")


class UserPreference(PjskBase):
    __tablename__ = "user_preferences"
    im_id = Column(String(30), primary_key=True)
    platform = Column(String(20), primary_key=True)
    option = Column(String(50), primary_key=True)
    value = Column(String(50), nullable=False)


class Alias(PjskBase):
    __tablename__ = "aliases"
    id = Column(BigInteger, primary_key=True, autoincrement=True)
    alias_type = Column(String(20), nullable=False)  # e.g., "music", "character"
    alias_type_id = Column(Integer, nullable=False)
    alias = Column(String(100), nullable=False)


class PendingAlias(PjskBase):
    __tablename__ = "pending_aliases"
    id = Column(BigInteger, primary_key=True, autoincrement=True)
    alias_type = Column(String(20), nullable=False)
    alias_type_id = Column(Integer, nullable=False)
    alias = Column(String(100), nullable=False)
    submitted_by = Column(String(100), nullable=False)
    submitted_at = Column(DateTime, nullable=False)


class RejectedAlias(PjskBase):
    __tablename__ = "rejected_aliases"
    id = Column(BigInteger, primary_key=True)
    alias_type = Column(String(20), nullable=False)
    alias_type_id = Column(Integer, nullable=False)
    alias = Column(String(100), nullable=False)
    reviewed_by = Column(String(100), nullable=False)
    reason = Column(String(255), nullable=False)
    reviewed_at = Column(DateTime, nullable=False)


class GroupAlias(PjskBase):
    __tablename__ = "group_aliases"
    group_id = Column(String(50), nullable=False, primary_key=True)
    alias_type = Column(String(20), nullable=False, primary_key=True)
    alias_type_id = Column(Integer, nullable=False, primary_key=True)
    alias = Column(String(100), nullable=False, primary_key=True)


class AliasAdmin(PjskBase):
    __tablename__ = "alias_admins"
    im_id = Column(String(100), primary_key=True)
    name = Column(String(100), nullable=False)

File: database/table/pjsk.go
package table

import "time"

type UserBinding struct {
	ID       int    `gorm:"column:id;primaryKey;autoIncrement"`
	Platform string `gorm:"column:platform;type:varchar(20);not null;uniqueIndex:uq_user_binding"`
	ImID     string `gorm:"column:im_id;type:varchar(30);not null;index;uniqueIndex:uq_user_binding"`
	UserID   string `gorm:"column:user_id;type:varchar(30);not null;uniqueIndex:uq_user_binding"`
	Server   string `gorm:"column:server;type:varchar(2);not null;uniqueIndex:uq_user_binding"`
	Visible  bool   `gorm:"column:visible;default:true"`

	DefaultRefs []UserDefaultBinding `gorm:"foreignKey:BindingID;constraint:OnDelete:CASCADE"`
}

func (UserBinding) TableName() string {
	return "user_bindings"
}

type UserDefaultBinding struct {
	ID        int    `gorm:"column:id;primaryKey;autoIncrement"`
	ImID      string `gorm:"column:im_id;type:varchar(30);not null;uniqueIndex:uq_user_default_binding"`
	Platform  string `gorm:"column:platform;type:varchar(20);not null;uniqueIndex:uq_user_default_binding"`
	Server    string `gorm:"column:server;type:varchar(7);not null;uniqueIndex:uq_user_default_binding"`
	BindingID int    `gorm:"column:binding_id;not null;index"`

	Binding UserBinding `gorm:"foreignKey:BindingID;references:ID;constraint:OnDelete:CASCADE"`
}

func (UserDefaultBinding) TableName() string {
	return "user_default_bindings"
}

type UserPreference struct {
	ImID     string `gorm:"column:im_id;type:varchar(30);primaryKey"`
	Platform string `gorm:"column:platform;type:varchar(20);primaryKey"`
	Option   string `gorm:"column:option;type:varchar(50);primaryKey"`
	Value    string `gorm:"column:value;type:varchar(50);not null"`
}

func (UserPreference) TableName() string {
	return "user_preferences"
}

type Alias struct {
	ID          int64  `gorm:"column:id;primaryKey;autoIncrement"`
	AliasType   string `gorm:"column:alias_type;type:varchar(20);not null"`
	AliasTypeID int    `gorm:"column:alias_type_id;not null"`
	Alias       string `gorm:"column:alias;type:varchar(100);not null"`
}

func (Alias) TableName() string {
	return "aliases"
}

type PendingAlias struct {
	ID          int64     `gorm:"column:id;primaryKey;autoIncrement"`
	AliasType   string    `gorm:"column:alias_type;type:varchar(20);not null"`
	AliasTypeID int       `gorm:"column:alias_type_id;not null"`
	Alias       string    `gorm:"column:alias;type:varchar(100);not null"`
	SubmittedBy string    `gorm:"column:submitted_by;type:varchar(100);not null"`
	SubmittedAt time.Time `gorm:"column:submitted_at;not null"`
}

func (PendingAlias) TableName() string {
	return "pending_aliases"
}

type RejectedAlias struct {
	ID          int64     `gorm:"column:id;primaryKey"`
	AliasType   string    `gorm:"column:alias_type;type:varchar(20);not null"`
	AliasTypeID int       `gorm:"column:alias_type_id;not null"`
	Alias       string    `gorm:"column:alias;type:varchar(100);not null"`
	ReviewedBy  string    `gorm:"column:reviewed_by;type:varchar(100);not null"`
	Reason      string    `gorm:"column:reason;type:varchar(255);not null"`
	ReviewedAt  time.Time `gorm:"column:reviewed_at;not null"`
}

func (RejectedAlias) TableName() string {
	return "rejected_aliases"
}

type GroupAlias struct {
	GroupID     string `gorm:"column:group_id;type:varchar(50);primaryKey"`
	AliasType   string `gorm:"column:alias_type;type:varchar(20);primaryKey"`
	AliasTypeID int    `gorm:"column:alias_type_id;primaryKey"`
	Alias       string `gorm:"column:alias;type:varchar(100);primaryKey"`
}

func (GroupAlias) TableName() string {
	return "group_aliases"
}

type AliasAdmin struct {
	ImID string `gorm:"column:im_id;type:varchar(100);primaryKey"`
	Name string `gorm:"column:name;type:varchar(100);not null"`
}

func (AliasAdmin) TableName() string {
	return "alias_admins"
}
